feat(linter)!: report unmatched rules with error exit code

camchenry opened the pull request:
- closes https://github.com/oxc-project/oxc/issues/6988

we now return an error exit code when there are unmatched rules. previously, we would print an error to stderr and continue running. however, this masked errors in some tests that actually had unmatched rules in them. these test cases now trigger a panic (in tests only, not at runtime), and help ensure that we are reporting an error message to the user for unknown rules, which we did not have any tests cases for before.

- fixes https://github.com/oxc-project/oxc/issues/7025

this also fixes https://github.com/oxc-project/oxc/issues/7025, where we were reporting rules as unmatched simply because they had been disabled prior to being configured. similar to https://github.com/oxc-project/oxc/issues/7009.

camchenry commented:
* **#7027** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7027?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camchenry and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac commented on crates/oxc_linter/src/builder.rs:
```diff
@@ -21,6 +21,13 @@ pub struct LinterBuilder {
 }
 
+/// An error that can occur while building a [`Linter`] from an [`Oxlintrc`].
+#[derive(Debug, Clone)]
```
Should implement std::error::Error for easy conversion to miette errors

DonIsaac reviewed (commented):
I need to review this more thoroughly. I'll keep looking when I have time (tonight or tomorrow)

Boshen commented on crates/oxc_language_server/src/main.rs:
```diff
@@ -359,6 +359,8 @@ impl Backend {
                     Oxlintrc::from_file(&config_path)
                         .expect("should have initialized linter with new options"),
                 )
+                // FIXME: Handle this error more gracefully and report it properly
```
Let's not crash the IDE 😅 shall we silence the error before the fix?

camchenry replied:
I think it's probably okay for now, since we already have two other unwrap/expect calls in this function? I need more expertise on the editor side though, I feel like the "right" way to handle this is to create an error popup, but maybe that happens automatically?

Boshen replied:
Yeah the correct way is to show a popup, I don't know what happens when it panics.

camchenry replied:
I can test it out and see what happens

nrayburn-tech replied:
From memory, I believe this is the behavior for VS Code and IntelliJ.
The language server will crash. If it crashes multiple times, a pop up will display indicating it has been disabled. You can restart your IDE to reenable it. You can check your logs to see what might have happened. (Oxc channel in VS Code or idea.log for IntelliJ)

If you want a popup, that needs to be communicated by the language server with a message to the IDE. I don’t know if a standard message exists, or if this would be a custom one.

For now, if you just catch the error and log it to stderr then I can fix the editor support after.

Boshen commented on apps/oxlint/src/lint.rs:
```diff
@@ -483,18 +501,20 @@ mod test {
+    // Previously, this test would pass and the unmatched rule would be ignored, but now we report that
+    // there was unmatched rule, because the typescript plugin has been disabled and we are trying to configure it.
     #[test]
+    #[should_panic(
```
Instead of testing for a panic that's coming out of the `test` function, shall we add another `test_failure` function instead.

Boshen commented on apps/oxlint/src/lint.rs:
```diff
@@ -119,9 +119,24 @@ impl Runner for LintRunner {
+                        OxcDiagnostic::warn(format!(
+                        "The following rules do not match the currently supported rules:\n{rules}"
+                    ))
+                        .with_help("Check that the plugin that contains this rule is enabled."),
```
This help message is somewhat confusing - it's either:

* we don't support the rule
* the rule is contained in a disabled plugin

camchenry replied:
maybe we should just skip this for now - I foresee in the future us being able to provide more helpful text here like finding which plugin contains this rule. for now though, we can just keep it simple

Boshen replied:
👍

Boshen commented:
### Merge activity

* **Oct 31, 11:26 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
feat(linter)!: report unmatched rules with error exit code (#7027)

- closes https://github.com/oxc-project/oxc/issues/6988

we now return an error exit code when there are unmatched rules. previously, we would print an error to stderr and continue running. however, this masked errors in some tests that actually had unmatched rules in them. these test cases now trigger a panic (in tests only, not at runtime), and help ensure that we are reporting an error message to the user for unknown rules, which we did not have any tests cases for before.

- fixes https://github.com/oxc-project/oxc/issues/7025

this also fixes https://github.com/oxc-project/oxc/issues/7025, where we were reporting rules as unmatched simply because they had been disabled prior to being configured. similar to https://github.com/oxc-project/oxc/issues/7009.

Boshen commented:
https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/11623823150/job/32371532541

We need to fix preact with this change 😅
